### PR TITLE
feat(donut-chart): add new pf-donut-chart component

### PR DESCRIPTION
**Pull request by sahil143**
New component based on [Patternfly Donut chart](http://www.patternfly.org/pattern-library/data-visualization/donut-chart/#/code).

https://rawgit.com/sahil143/patternfly-webcomponents/donut-chart-dist/app/app.html?dir=pf-donut-chart&file=index.html

**.eslintrc** changes:
`"dot-notation": [1, {"allowPattern": "^[a-z]+(_[a-z]+)+$"}]` because spread operator for objects is not working with gulp-eslint so simple assignment `obj["newkey"] = value` is used.

### CODE REVIEW

File: src/pf-donut-chart/index.html [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+  <pf-donut-chart id="donut" title="Utilization Chart" target-selector="#chart" columns="[
+                                                                                  ['Used', 60],
+                                                                                  ['Available', 40]
+                                                                                ]" data="{
+                                                                                  'groups': [
+                                                                                      ['used', 'available']
+                                                                                  ],
+                                                                                  'order': null
+                                                                                }" width="200" height "171"></pf-donut-chart>
```
Comment (cdcabrera) [inline src/pf-donut-chart/index.html]: Simple missing attribute "equals" for height

File: src/pf-donut-chart/index.html [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                                              'data3': '#ec7a08' 
+                                            }" json="[
+                                              {'name': 'www.site1.com', 'data1': 200, 'data2': 200, 'data3': 400},
+                                              {'name': 'www.site2.com', 'data1': 100, 'data2': 300, 'data3': 400}
+                                            ]" data="{ 
+                                              'keys': {
+                                              'value': ['data1', 'data2', 'data3'] 
+                                              }
+                                            }" title="donut2" width="200" height="171"></pf-donut-chart>
+  <pf-donut-chart id="donut3" title="donut3" target-selector="#chart3" url="https://gist.githubusercontent.com/sahil143/b0c56761599c64080a3a042ad779e097/raw/97c51c6c4db4a121b7f49e7f58cca675c8528058/test.json"></pf-donut-chart>
```
Comment (cdcabrera) [inline src/pf-donut-chart/index.html]: Is the plan to look at hosting this under the ```patternfly-webcomponents``` repository, or within a gist controlled by PatternFly?

File: src/pf-donut-chart/index.html [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  <pf-donut-chart id="donut4" target-selector="#chart4" url="https://gist.githubusercontent.com/sahil143/0a8dbe79028a673d4b87d7fdf1d0c098/raw/6e2f599e92c3bd804d0bff347804a0b12b994803/test.csv"></pf-donut-chart>
```
Comment (cdcabrera) [inline src/pf-donut-chart/index.html]: Same hosting question as above...

File: src/pf-donut-chart/index.html [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    // #donut Tooltip Example
+    donut.tooltip = {
+      show: true,
+      contents: function (d) {
+        return '<span class="donut-tooltip-pf" style="white-space: nowrap;">' +
+          (Math.round(d[0].ratio * 1000) / 10) + '%' + ' MHz ' + d[0].name +
+          '</span>';
+      }
+    }
```
Comment (cdcabrera) [inline src/pf-donut-chart/index.html]: Missing semicolon

File: src/pf-donut-chart/index.html [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    donut1.tooltip = {
+      show: true,
+      contents: tooltipContent
+    };
+
+    // #donut2
+    donut2.tooltip = {
+      show: true,
+      contents: tooltipContent
+    }
```
Comment (cdcabrera) [inline src/pf-donut-chart/index.html]: Semicolon

File: src/pf-donut-chart/index.html [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    // #donut3
+    donut3.legend = {
+      show: true
+    }
```
Comment (cdcabrera) [inline src/pf-donut-chart/index.html]: Semicolon

File: src/pf-donut-chart/index.html [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+    donut3.tooltip = {
+      show: true
+    }
```
Comment (cdcabrera) [inline src/pf-donut-chart/index.html]: Semicolon

File: src/pf-donut-chart/index.html [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+    donut3.colors = {
+      data1: '#0088ce',
+      data2: '#3f9c35',
+      data3: '#ec7a08',
+      data4: '#123455',
+      data6: '#abcdef'
+    }
```
Comment (cdcabrera) [inline src/pf-donut-chart/index.html]: Semicolon

File: src/pf-donut-chart/pf-donut-chart.component.js [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+  set height(value) {
+    if (this._height !== value) {
+      this._height = value;
+      this.setAttribute('height', value);
+    }
+  }
+
+  /**
+   * Get charts legend
+   */
```
Comment (cdcabrera) [inline src/pf-donut-chart/pf-donut-chart.component.js]: Probably haven't called this out before, but placing the "returns" in here rounds your annotations out nicely
```js
* Get charts legend
* @returns {string}
```

File: src/pf-donut-chart/pf-donut-chart.component.js [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+   * @param {object} value object contains legend properties
+   */
+  set legend(value) {
+    this._legend = value;
+    this._prepareChart();
+  }
+
+  /**
+   * Get chart title
+   */
```
Comment (cdcabrera) [inline src/pf-donut-chart/pf-donut-chart.component.js]: Placing "returns" comment, similar to ```* @returns {SOMETYPE}```

File: src/pf-donut-chart/pf-donut-chart.component.js [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+  set title(value) {
+    if (this._title !== value) {
+      this._title = value;
+      this.setAttribute('title', value);
+    }
+  }
+
+  /**
+   * Get target-selector
+   */
```
Comment (cdcabrera) [inline src/pf-donut-chart/pf-donut-chart.component.js]: Placing "returns" comment, similar to ```* @returns {SOMETYPE}```

File: src/pf-donut-chart/pf-donut-chart.component.js [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+  set targetSelector(value) {
+    if (this._targetSelector !== value) {
+      this._targetSelector = value;
+      this.setAttribute('target-selector', value);
+    }
+  }
+
+  /**
+   * get additional data
+   */
```
Comment (cdcabrera) [inline src/pf-donut-chart/pf-donut-chart.component.js]: Placing "returns" comment, similar to ```* @returns {SOMETYPE}```

File: src/pf-donut-chart/pf-donut-chart.component.js [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+   * @param {object} obj additional data object
+   */
+  set data(obj) {
+    this._additionalData = obj;
+    this._prepareChart();
+  }
+
+  /**
+   * get colors
+   */
```
Comment (cdcabrera) [inline src/pf-donut-chart/pf-donut-chart.component.js]: Placing "returns" comment, similar to ```* @returns {SOMETYPE}```

File: src/pf-donut-chart/pf-donut-chart.component.js [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+   * @param {object} obj colors object
+   */
+  set colors(obj) {
+    this._colors = obj;
+    this._prepareData();
+  }
+
+  /**
+   * get c3 tooltip object
+   */
```
Comment (cdcabrera) [inline src/pf-donut-chart/pf-donut-chart.component.js]: Placing "returns" comment, similar to ```* @returns {SOMETYPE}```

File: src/pf-donut-chart/index.html (hunk 2 above)
Comment (priley86) [inline src/pf-donut-chart/index.html]: Good catch...i'm not sure what the standard is on this.

**Comment by sahil143** [comment]
@cdcabrera  Yes, a configuration option for features that `pf-donut-chart` don't have right now would be great. But I think we should wait on spread operator for now because it will be really easy to integrate with the help of spread operator. On the other hand, If we want to do it now then we would have to use simple assignment or different attributes for options like `axis`, `grid` to integrate it into `_prepareChart()` and we will end up with a huge pile of attributes. Also, current `pf-donut-chart` consists of all options/feature which is showcased on patternfly donut chart test page. Thoughts on this?

Cc: @priley86

**Comment by priley86** [comment]
@sahil143 i completely agree about the spread operator (it would be a good candidate for this sort of data mapping). It appears that array spreads work in this repo, but object spreads do not currently. We can try add to support for the object spread in a subsequent PR.

File: src/pf-donut-chart/pf-donut-chart.component.js [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+  set width(value) {
+    if (this._width !== value) {
+      this._width = value;
+      this.setAttribute('width', value);
+    }
+  }
+
+  /**
+   *  Get chart height
+   * @returns {numbers}
```
Comment (sahil143) [inline src/pf-donut-chart/pf-donut-chart.component.js]: just a typo. I will change it with the unit tests.

**Comment by cdcabrera** [comment]
@sahil143 @priley86 per @bleathem recommendation lets look at placing the example JSON inside the repository

File: index.html [hunk 16]
```diff
@@ -63,6 +63,14 @@
           </li>
         </ul>
       </div>
+      <div class="col-sm-6">
+        <h2>Charts</h2>
+        <ul>
+          <li>
+            <a href="app/app.html?dir=pf-donut-chart&file=index.html">Pf-donut-chart</a>
+          </li>
```
Comment (priley86) [inline index.html]: One more thing... can we change the link text to "Donut Charts" here please?

**Comment by priley86** [comment]
this looks great @sahil143 ! I am happy w/ the changes. I'd like to merge this soon so that we might complete your GSoC reports. We can continue to iterate on charts in the future!

**Pull request merged**